### PR TITLE
Extend redirection SPI to UNION ALL types of redirections

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: core/trino-spi/src/main/java/io/trino/spi/connector/Redirection.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ */
+package io.trino.spi.connector;
+
+import io.trino.spi.predicate.TupleDomain;
+
+import java.util.Map;
+
+import static java.util.Objects.requireNonNull;
+
+public class Redirection
```
Comment (raunaqmorarka) [inline core/trino-spi/src/main/java/io/trino/spi/connector/Redirection.java:22]: Should this be an inner class of TableScanRedirectApplicationResult ?

**Review by raunaqmorarka — APPROVED** [review]

**Comment by erichwang** [comment]
@sopel39, could we set up some time to discuss this change? Probably want to invite @martint too. This is mainly because I'm a bit wary about encoding the concept of table redirects into the external facing SPI. There's probably a more generalized form we need to consider here.

File: core/trino-spi/src/main/java/io/trino/spi/connector/Redirection.java (hunk 1 above)
Comment (martint) [reply core/trino-spi/src/main/java/io/trino/spi/connector/Redirection.java:22]: Yes, let's make it an inner class. The concept of "Redirection" as a first class entity doesn't make sense -- this is just a table + filter. "Redirection" is just the way such an entity is used in the context of the redirection API

Comment (sopel39) [reply core/trino-spi/src/main/java/io/trino/spi/connector/Redirection.java]: made this inner class

**Review by erichwang — APPROVED** [review]
The PR looks fine by itself. One thing we would want to think about here is how UNION ALL as an engineering direction might impact further developments for things like incremental loading (which is a hard problem). I suspect there will be some tension between the 2. But it's not something that necessarily blocks this work at this point.

**Review by martint — APPROVED** [review]

**Pull request merged**